Adopt Safer CPP in CachedFrame.cpp

jelee53 opened the pull request:
#### 6fd821c4fc29440734ca7ef412da649a6e54d553
<pre>
Adopt Safer CPP in CachedFrame.cpp
<a href="https://rdar.apple.com/166737805">rdar://166737805</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=304359">https://bugs.webkit.org/show_bug.cgi?id=304359</a>

Reviewed by Ryosuke Niwa.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.</a>

No new tests needed.

* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/history/CachedFrame.cpp:
(WebCore::CachedFrameBase::restore):
(WebCore::CachedFrame::CachedFrame):
(WebCore::CachedFrame::destroy):
(WebCore::CachedFrame::hasInsecureContent const):

Canonical link: <a href="https://commits.webkit.org/304778@main">https://commits.webkit.org/304778@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7a74abbec74388fafac35bd97aaa51a1362a3767

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136485 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8842 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47765 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144198 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ab066f85-9b91-4f02-a689-ca1e69afd410) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138357 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9529 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8686 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104370 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/62533e22-1f4c-426a-9777-34924ccf13ec) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139430 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6954 "Found 1 new test failure: fast/forms/datalist/datalist-crash-when-dynamic.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122296 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85205 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6e96077e-6307-4a9b-be21-a5b95b6d80a2) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6598 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4260 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4791 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115900 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40500 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146947 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8524 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41069 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112710 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8541 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7164 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113054 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6534 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118604 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/62513 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21049 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8572 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36656 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8291 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72131 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8512 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8364 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/history/CachedFrame.cpp:
```diff
@@ -280,8 +281,9 @@ void CachedFrame::destroy()
-    if (!m_isMainFrame && m_view->frame().page()) {
+    RefPtr view = m_view;
+    Ref frame = view->frame();
+    if (!m_isMainFrame && view->frame().page()) {
```
Existing issue, but still good to fix:
```suggestion
    if (!m_isMainFrame && frame->page()) {
```

annevk commented on Source/WebCore/history/CachedFrame.cpp:
```diff
@@ -280,8 +281,9 @@ void CachedFrame::destroy()
+    RefPtr view = m_view;
+    Ref frame = view->frame();
+    if (!m_isMainFrame && view->frame().page()) {
         if (RefPtr localFrame = dynamicDowncast<LocalFrame>(frame.get()))
```
```suggestion
        if (RefPtr localFrame = dynamicDowncast<LocalFrame>(frame))
```

annevk commented on Source/WebCore/history/CachedFrame.cpp:
```diff
@@ -293,7 +295,7 @@ void CachedFrame::destroy()
         m_cachedFramePlatformData->clear();
 
-    if (RefPtr localFrameView = dynamicDowncast<LocalFrameView>(m_view.get()); localFrameView)
+    if (RefPtr localFrameView = dynamicDowncast<LocalFrameView>(view.get()); localFrameView)
```
```suggestion
    if (RefPtr localFrameView = dynamicDowncast<LocalFrameView>(view); localFrameView)
```

annevk commented on Source/WebCore/history/CachedFrame.cpp:
```diff
@@ -173,7 +173,7 @@ CachedFrame::CachedFrame(Frame& frame)
     if (document) {
         RELEASE_ASSERT(document->window());
-        RELEASE_ASSERT(document->window()->frame());
+        RELEASE_ASSERT(document->protectedWindow()->protectedFrame());
```
Are you sure we need `protectedFrame()` here? Also, given it's a `RELEASE_ASSERT` it makes sense to put `document->window()` in a local variable as we use it twice.

jelee53 replied:
I'm actually not sure! Let me remove it and try it out.

annevk commented on Source/WebCore/history/CachedFrame.cpp:
```diff
@@ -342,7 +345,7 @@ UsedLegacyTLS CachedFrame::usedLegacyTLS() const
 HasInsecureContent CachedFrame::hasInsecureContent() const
 {
-    if (auto* document = this->document()) {
+    if (CheckedPtr document = this->document()) {
```
This should be RefPtr.

jelee53 committed:
Adopt Safer CPP in CachedFrame.cpp
rdar://166737805
https://bugs.webkit.org/show_bug.cgi?id=304359

Reviewed by Ryosuke Niwa.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.

No new tests needed.

* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/history/CachedFrame.cpp:
(WebCore::CachedFrameBase::restore):
(WebCore::CachedFrame::CachedFrame):
(WebCore::CachedFrame::destroy):
(WebCore::CachedFrame::hasInsecureContent const):

Canonical link: https://commits.webkit.org/304778@main
